// ng-zorro-antd/bundles/ng-zorro-antd-cdk-overflow.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ng-zorro-antd/cdk/resize-observer'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/cdk/overflow', ['exports', '@angular/core', 'ng-zorro-antd/cdk/resize-observer', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd'].cdk = global['ng-zorro-antd'].cdk || {}, global['ng-zorro-antd'].cdk.overflow = {}), global.ng.core, global['ng-zorro-antd'].cdk['resize-observer'], global.rxjs, global.rxjs.operators));
}(this, (function (exports, core, resizeObserver, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var NzOverflowItemDirective = /** @class */ (function () {
        function NzOverflowItemDirective(nzResizeObserver, elementRef, cdr) {
            var _this = this;
            this.nzResizeObserver = nzResizeObserver;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.overflowStyle = undefined;
            this.itemWidth$ = this.nzResizeObserver.observe(this.elementRef.nativeElement).pipe(operators.map(function (_a) {
                var _b = __read(_a, 1), item = _b[0];
                return item.target.offsetWidth;
            }), operators.distinctUntilChanged(), operators.startWith(undefined), operators.tap(function (width) {
                _this.itemWidth = width;
            }));
            this.itemWidth = undefined;
        }
        NzOverflowItemDirective.prototype.setItemStyle = function (display, order) {
            var mergedHidden = !display;
            this.overflowStyle = {
                opacity: mergedHidden ? 0 : 1,
                height: mergedHidden ? 0 : undefined,
                overflowY: mergedHidden ? 'hidden' : undefined,
                order: order,
                pointerEvents: mergedHidden ? 'none' : undefined,
                position: mergedHidden ? 'absolute' : undefined
            };
            this.cdr.detectChanges();
        };
        return NzOverflowItemDirective;
    }());
    NzOverflowItemDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[nzOverflowItem]',
                    host: {
                        '[style]': 'overflowStyle'
                    }
                },] }
    ];
    NzOverflowItemDirective.ctorParameters = function () { return [
        { type: resizeObserver.NzResizeObserver },
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };

    var NzOverflowRestDirective = /** @class */ (function () {
        function NzOverflowRestDirective(nzResizeObserver, elementRef, cdr) {
            var _this = this;
            this.nzResizeObserver = nzResizeObserver;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.restStyle = undefined;
            this.restWidth$ = this.nzResizeObserver.observe(this.elementRef.nativeElement).pipe(operators.map(function (_a) {
                var _b = __read(_a, 1), item = _b[0];
                return item.target.offsetWidth;
            }), operators.startWith(0), operators.tap(function (width) { return (_this.restWidth = width); }));
            this.restWidth = 0;
        }
        NzOverflowRestDirective.prototype.setRestStyle = function (display, order) {
            var mergedHidden = !display;
            this.restStyle = {
                opacity: mergedHidden ? 0 : 1,
                height: mergedHidden ? 0 : undefined,
                overflowY: mergedHidden ? 'hidden' : undefined,
                order: order,
                pointerEvents: mergedHidden ? 'none' : undefined,
                position: mergedHidden ? 'absolute' : undefined
            };
            this.cdr.detectChanges();
        };
        return NzOverflowRestDirective;
    }());
    NzOverflowRestDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[nzOverflowRest]',
                    host: {
                        '[style]': 'restStyle'
                    }
                },] }
    ];
    NzOverflowRestDirective.ctorParameters = function () { return [
        { type: resizeObserver.NzResizeObserver },
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };

    var NzOverflowSuffixDirective = /** @class */ (function () {
        function NzOverflowSuffixDirective(nzResizeObserver, elementRef, cdr) {
            var _this = this;
            this.nzResizeObserver = nzResizeObserver;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.suffixStyle = {};
            this.suffixWidth$ = this.nzResizeObserver.observe(this.elementRef.nativeElement).pipe(operators.map(function (_a) {
                var _b = __read(_a, 1), item = _b[0];
                return item.target.offsetWidth;
            }), operators.tap(function (width) { return (_this.suffixWidth = width); }));
            this.suffixWidth = 0;
        }
        NzOverflowSuffixDirective.prototype.setSuffixStyle = function (start, order) {
            if (start !== null) {
                this.suffixStyle = {
                    position: 'absolute',
                    left: start + "px",
                    top: 0,
                    opacity: 1,
                    order: order
                };
            }
            else {
                this.suffixStyle = {
                    opacity: 1,
                    order: order
                };
            }
            this.cdr.detectChanges();
        };
        return NzOverflowSuffixDirective;
    }());
    NzOverflowSuffixDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[nzOverflowSuffix]',
                    host: {
                        '[style]': 'suffixStyle'
                    }
                },] }
    ];
    NzOverflowSuffixDirective.ctorParameters = function () { return [
        { type: resizeObserver.NzResizeObserver },
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };

    var NzOverflowContainerComponent = /** @class */ (function () {
        function NzOverflowContainerComponent(nzResizeObserver, elementRef, cdr) {
            this.nzResizeObserver = nzResizeObserver;
            this.elementRef = elementRef;
            this.cdr = cdr;
            this.contentInit$ = new rxjs.Subject();
            this.overflowItems = undefined;
            this.overflowSuffix = undefined;
            this.overflowRest = undefined;
            this.overflowItems$ = new rxjs.ReplaySubject(1);
            this.destroy$ = new rxjs.Subject();
            this.containerWidth$ = this.nzResizeObserver
                .observe(this.elementRef.nativeElement)
                .pipe(operators.map(function (_d) {
                var _e = __read(_d, 1), item = _e[0];
                return item.target.clientWidth || 0;
            }));
            this.restWidth$ = new rxjs.BehaviorSubject(0);
            this.suffixWidth$ = new rxjs.BehaviorSubject(0);
            this.suffixFixedStart$ = new rxjs.BehaviorSubject(null);
            this.displayCount$ = new rxjs.BehaviorSubject(Number.MAX_SAFE_INTEGER);
            this.restReady$ = new rxjs.BehaviorSubject(false);
            this.maxRestWith$ = this.restWidth$.pipe(operators.pairwise(), operators.map(function (_d) {
                var _e = __read(_d, 2), prevRestWidth = _e[0], restWidth = _e[1];
                return Math.max(prevRestWidth, restWidth);
            }));
            this.omittedItems$ = rxjs.combineLatest([this.overflowItems$, this.displayCount$]).pipe(operators.withLatestFrom(this.contentInit$), operators.map(function (_d) {
                var _e = __read(_d, 1), _f = __read(_e[0], 2), overflowItems = _f[0], displayCount = _f[1];
                return overflowItems.toArray().slice(displayCount + 1);
            }));
            this.displayRest$ = rxjs.combineLatest([this.restReady$, this.omittedItems$]).pipe(operators.map(function (_d) {
                var _e = __read(_d, 2), restReady = _e[0], omittedItems = _e[1];
                return restReady && !!omittedItems.length;
            }));
        }
        NzOverflowContainerComponent.prototype.updateDisplayCount = function (count, notReady) {
            this.displayCount$.next(count);
            if (this.overflowItems && !notReady) {
                this.restReady$.next(count < this.overflowItems.length - 1);
            }
        };
        NzOverflowContainerComponent.prototype.ngOnInit = function () {
            var _this = this;
            var overflowItemsWidth$ = this.overflowItems$.pipe(operators.switchMap(function (items) { return rxjs.combineLatest(items.map(function (item) { return item.itemWidth$; })); }));
            this.overflowItems$.pipe(operators.takeUntil(this.destroy$)).subscribe(function (overflowItems) {
                if (!overflowItems.length) {
                    _this.displayCount$.next(0);
                    _this.suffixFixedStart$.next(null);
                }
            });
            rxjs.combineLatest([overflowItemsWidth$, this.containerWidth$, this.maxRestWith$, this.restWidth$, this.suffixWidth$])
                .pipe(operators.filter(function (_d) {
                var _e = __read(_d, 3), containerWidth = _e[1], maxRestWith = _e[2];
                return !!(containerWidth && maxRestWith);
            }), operators.takeUntil(this.destroy$))
                .subscribe(function (_d) {
                var _e = __read(_d, 5), overflowItemsWidth = _e[0], containerWidth = _e[1], maxRestWith = _e[2], restWidth = _e[3], suffixWidth = _e[4];
                var totalWidth = suffixWidth;
                var len = overflowItemsWidth.length;
                var lastIndex = len - 1;
                for (var i = 0; i < len; i += 1) {
                    var currentItemWidth = overflowItemsWidth[i];
                    // Break since data not ready
                    if (currentItemWidth === undefined) {
                        _this.updateDisplayCount(i - 1, true);
                        break;
                    }
                    else {
                        // Find best match
                        totalWidth += currentItemWidth;
                        if (
                        // Only one means `totalWidth` is the final width
                        (lastIndex === 0 && totalWidth <= containerWidth) ||
                            // Last two width will be the final width
                            (i === lastIndex - 1 &&
                                overflowItemsWidth[lastIndex] !== undefined &&
                                totalWidth + overflowItemsWidth[lastIndex] <= containerWidth)) {
                            // Additional check if match the end
                            _this.updateDisplayCount(lastIndex);
                            _this.suffixFixedStart$.next(null);
                            break;
                        }
                        else if (totalWidth + maxRestWith > containerWidth) {
                            // Can not hold all the content to show rest
                            _this.updateDisplayCount(i - 1);
                            _this.suffixFixedStart$.next(totalWidth - currentItemWidth - suffixWidth + restWidth);
                            break;
                        }
                        _this.cdr.detectChanges();
                    }
                }
                if (_this.overflowSuffix &&
                    overflowItemsWidth[0] !== undefined &&
                    overflowItemsWidth[0] + suffixWidth > containerWidth) {
                    _this.suffixFixedStart$.next(null);
                }
                _this.cdr.detectChanges();
            });
            rxjs.combineLatest([this.suffixFixedStart$, this.displayCount$])
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (_d) {
                var _e = __read(_d, 2), suffixFixedStart = _e[0], displayCount = _e[1];
                var _a;
                (_a = _this.overflowSuffix) === null || _a === void 0 ? void 0 : _a.setSuffixStyle(suffixFixedStart, displayCount);
            });
            rxjs.combineLatest([this.displayCount$, this.overflowItems$])
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (_d) {
                var _e = __read(_d, 2), displayCount = _e[0], overflowItems = _e[1];
                return overflowItems.forEach(function (item, index) { return item.setItemStyle(index <= displayCount, index); });
            });
            rxjs.combineLatest([this.displayRest$, this.displayCount$])
                .pipe(operators.takeUntil(this.destroy$))
                .subscribe(function (_d) {
                var _e = __read(_d, 2), displayRest = _e[0], displayCount = _e[1];
                var _a;
                (_a = _this.overflowRest) === null || _a === void 0 ? void 0 : _a.setRestStyle(displayRest, displayRest ? displayCount : Number.MAX_SAFE_INTEGER);
            });
        };
        NzOverflowContainerComponent.prototype.ngAfterContentInit = function () {
            var _a, _b, _c;
            (_a = this.overflowItems) === null || _a === void 0 ? void 0 : _a.changes.pipe(operators.startWith(this.overflowItems)).subscribe(this.overflowItems$);
            (_b = this.overflowSuffix) === null || _b === void 0 ? void 0 : _b.suffixWidth$.subscribe(this.suffixWidth$);
            (_c = this.overflowRest) === null || _c === void 0 ? void 0 : _c.restWidth$.subscribe(this.restWidth$);
            this.contentInit$.next();
        };
        NzOverflowContainerComponent.prototype.ngOnDestroy = function () {
            this.destroy$.next();
            this.destroy$.complete();
        };
        return NzOverflowContainerComponent;
    }());
    NzOverflowContainerComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'nz-overflow-container',
                    template: " <ng-content></ng-content>\n    <ng-content select=\"[appOverflowRest]\"></ng-content>\n    <ng-content select=\"[appOverflowSuffix]\"></ng-content>",
                    providers: [resizeObserver.NzResizeObserver],
                    changeDetection: core.ChangeDetectionStrategy.OnPush
                },] }
    ];
    NzOverflowContainerComponent.ctorParameters = function () { return [
        { type: resizeObserver.NzResizeObserver },
        { type: core.ElementRef },
        { type: core.ChangeDetectorRef }
    ]; };
    NzOverflowContainerComponent.propDecorators = {
        overflowItems: [{ type: core.ContentChildren, args: [NzOverflowItemDirective,] }],
        overflowSuffix: [{ type: core.ContentChild, args: [NzOverflowSuffixDirective,] }],
        overflowRest: [{ type: core.ContentChild, args: [NzOverflowRestDirective,] }]
    };

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */
    var NzOverflowModule = /** @class */ (function () {
        function NzOverflowModule() {
        }
        return NzOverflowModule;
    }());
    NzOverflowModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [resizeObserver.NzResizeObserverModule],
                    declarations: [
                        NzOverflowContainerComponent,
                        NzOverflowItemDirective,
                        NzOverflowRestDirective,
                        NzOverflowSuffixDirective
                    ],
                    exports: [NzOverflowContainerComponent, NzOverflowItemDirective, NzOverflowRestDirective, NzOverflowSuffixDirective]
                },] }
    ];

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NzOverflowContainerComponent = NzOverflowContainerComponent;
    exports.NzOverflowItemDirective = NzOverflowItemDirective;
    exports.NzOverflowModule = NzOverflowModule;
    exports.NzOverflowRestDirective = NzOverflowRestDirective;
    exports.NzOverflowSuffixDirective = NzOverflowSuffixDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
