// ng-zorro-antd/bundles/ng-zorro-antd-core-trans-button.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/core/trans-button', ['exports', '@angular/common', '@angular/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd'].core = global['ng-zorro-antd'].core || {}, global['ng-zorro-antd'].core['trans-button'] = {}), global.ng.common, global.ng.core));
}(this, (function (exports, common, core) { 'use strict';

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */
    var NzTransButtonDirective = /** @class */ (function () {
        function NzTransButtonDirective() {
        }
        return NzTransButtonDirective;
    }());
    NzTransButtonDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: 'button[nz-trans-button]',
                    host: {
                        '[style.border]': '"0"',
                        '[style.background]': '"transparent"',
                        '[style.padding]': '"0"',
                        '[style.line-height]': '"inherit"'
                    }
                },] }
    ];

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */
    var NzTransButtonModule = /** @class */ (function () {
        function NzTransButtonModule() {
        }
        return NzTransButtonModule;
    }());
    NzTransButtonModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [NzTransButtonDirective],
                    exports: [NzTransButtonDirective],
                    imports: [common.CommonModule]
                },] }
    ];

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NzTransButtonDirective = NzTransButtonDirective;
    exports.NzTransButtonModule = NzTransButtonModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
