// ng-zorro-antd/bundles/ng-zorro-antd-popconfirm.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/cdk/bidi'), require('@angular/common'), require('@angular/core'), require('rxjs'), require('rxjs/operators'), require('ng-zorro-antd/core/animation'), require('ng-zorro-antd/core/config'), require('ng-zorro-antd/core/no-animation'), require('ng-zorro-antd/core/util'), require('ng-zorro-antd/tooltip'), require('@angular/cdk/a11y'), require('@angular/cdk/overlay'), require('ng-zorro-antd/button'), require('ng-zorro-antd/core/outlet'), require('ng-zorro-antd/core/overlay'), require('ng-zorro-antd/i18n'), require('ng-zorro-antd/icon')) :
    typeof define === 'function' && define.amd ? define('ng-zorro-antd/popconfirm', ['exports', '@angular/cdk/bidi', '@angular/common', '@angular/core', 'rxjs', 'rxjs/operators', 'ng-zorro-antd/core/animation', 'ng-zorro-antd/core/config', 'ng-zorro-antd/core/no-animation', 'ng-zorro-antd/core/util', 'ng-zorro-antd/tooltip', '@angular/cdk/a11y', '@angular/cdk/overlay', 'ng-zorro-antd/button', 'ng-zorro-antd/core/outlet', 'ng-zorro-antd/core/overlay', 'ng-zorro-antd/i18n', 'ng-zorro-antd/icon'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['ng-zorro-antd'] = global['ng-zorro-antd'] || {}, global['ng-zorro-antd'].popconfirm = {}), global.ng.cdk.bidi, global.ng.common, global.ng.core, global.rxjs, global.rxjs.operators, global['ng-zorro-antd'].core.animation, global['ng-zorro-antd'].core.config, global['ng-zorro-antd'].core['no-animation'], global['ng-zorro-antd'].core.util, global['ng-zorro-antd'].tooltip, global.ng.cdk.a11y, global.ng.cdk.overlay, global['ng-zorro-antd'].button, global['ng-zorro-antd'].core.outlet, global['ng-zorro-antd'].core.overlay, global['ng-zorro-antd'].i18n, global['ng-zorro-antd'].icon));
}(this, (function (exports, bidi, common, core, rxjs, operators, animation, config, noAnimation, util, tooltip, a11y, overlay, button, outlet, overlay$1, i18n, icon) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from, pack) {
        if (pack || arguments.length === 2)
            for (var i = 0, l = from.length, ar; i < l; i++) {
                if (ar || !(i in from)) {
                    if (!ar)
                        ar = Array.prototype.slice.call(from, 0, i);
                    ar[i] = from[i];
                }
            }
        return to.concat(ar || from);
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var NZ_CONFIG_MODULE_NAME = 'popconfirm';
    var NzPopconfirmDirective = /** @class */ (function (_super) {
        __extends(NzPopconfirmDirective, _super);
        function NzPopconfirmDirective(elementRef, hostView, resolver, renderer, noAnimation, nzConfigService) {
            var _this = _super.call(this, elementRef, hostView, resolver, renderer, noAnimation, nzConfigService) || this;
            _this._nzModuleName = NZ_CONFIG_MODULE_NAME;
            _this.trigger = 'click';
            _this.placement = 'top';
            _this.nzCondition = false;
            _this.nzPopconfirmShowArrow = true;
            _this.nzPopconfirmBackdrop = false;
            _this.nzAutofocus = null;
            // eslint-disable-next-line @angular-eslint/no-output-rename
            _this.visibleChange = new core.EventEmitter();
            _this.nzOnCancel = new core.EventEmitter();
            _this.nzOnConfirm = new core.EventEmitter();
            _this.componentFactory = _this.resolver.resolveComponentFactory(NzPopconfirmComponent);
            return _this;
        }
        NzPopconfirmDirective.prototype.getProxyPropertyMap = function () {
            var _this = this;
            return Object.assign({ nzOkText: ['nzOkText', function () { return _this.nzOkText; }], nzOkType: ['nzOkType', function () { return _this.nzOkType; }], nzCancelText: ['nzCancelText', function () { return _this.nzCancelText; }], nzCondition: ['nzCondition', function () { return _this.nzCondition; }], nzIcon: ['nzIcon', function () { return _this.nzIcon; }], nzPopconfirmShowArrow: ['nzPopconfirmShowArrow', function () { return _this.nzPopconfirmShowArrow; }], nzPopconfirmBackdrop: ['nzBackdrop', function () { return _this.nzPopconfirmBackdrop; }], nzAutoFocus: ['nzAutoFocus', function () { return _this.nzAutofocus; }] }, _super.prototype.getProxyPropertyMap.call(this));
        };
        /**
         * @override
         */
        NzPopconfirmDirective.prototype.createComponent = function () {
            var _this = this;
            _super.prototype.createComponent.call(this);
            this.component.nzOnCancel.pipe(operators.takeUntil(this.destroy$)).subscribe(function () {
                _this.nzOnCancel.emit();
            });
            this.component.nzOnConfirm.pipe(operators.takeUntil(this.destroy$)).subscribe(function () {
                _this.nzOnConfirm.emit();
            });
        };
        return NzPopconfirmDirective;
    }(tooltip.NzTooltipBaseDirective));
    NzPopconfirmDirective.decorators = [
        { type: core.Directive, args: [{
                    selector: '[nz-popconfirm]',
                    exportAs: 'nzPopconfirm',
                    host: {
                        '[class.ant-popover-open]': 'visible'
                    }
                },] }
    ];
    NzPopconfirmDirective.ctorParameters = function () { return [
        { type: core.ElementRef },
        { type: core.ViewContainerRef },
        { type: core.ComponentFactoryResolver },
        { type: core.Renderer2 },
        { type: noAnimation.NzNoAnimationDirective, decorators: [{ type: core.Host }, { type: core.Optional }] },
        { type: config.NzConfigService }
    ]; };
    NzPopconfirmDirective.propDecorators = {
        title: [{ type: core.Input, args: ['nzPopconfirmTitle',] }],
        directiveTitle: [{ type: core.Input, args: ['nz-popconfirm',] }],
        trigger: [{ type: core.Input, args: ['nzPopconfirmTrigger',] }],
        placement: [{ type: core.Input, args: ['nzPopconfirmPlacement',] }],
        origin: [{ type: core.Input, args: ['nzPopconfirmOrigin',] }],
        mouseEnterDelay: [{ type: core.Input, args: ['nzPopconfirmMouseEnterDelay',] }],
        mouseLeaveDelay: [{ type: core.Input, args: ['nzPopconfirmMouseLeaveDelay',] }],
        overlayClassName: [{ type: core.Input, args: ['nzPopconfirmOverlayClassName',] }],
        overlayStyle: [{ type: core.Input, args: ['nzPopconfirmOverlayStyle',] }],
        visible: [{ type: core.Input, args: ['nzPopconfirmVisible',] }],
        nzOkText: [{ type: core.Input }],
        nzOkType: [{ type: core.Input }],
        nzCancelText: [{ type: core.Input }],
        nzIcon: [{ type: core.Input }],
        nzCondition: [{ type: core.Input }],
        nzPopconfirmShowArrow: [{ type: core.Input }],
        nzPopconfirmBackdrop: [{ type: core.Input }],
        nzAutofocus: [{ type: core.Input }],
        visibleChange: [{ type: core.Output, args: ['nzPopconfirmVisibleChange',] }],
        nzOnCancel: [{ type: core.Output }],
        nzOnConfirm: [{ type: core.Output }]
    };
    __decorate([
        util.InputBoolean()
    ], NzPopconfirmDirective.prototype, "nzCondition", void 0);
    __decorate([
        util.InputBoolean()
    ], NzPopconfirmDirective.prototype, "nzPopconfirmShowArrow", void 0);
    __decorate([
        config.WithConfig()
    ], NzPopconfirmDirective.prototype, "nzPopconfirmBackdrop", void 0);
    __decorate([
        config.WithConfig()
    ], NzPopconfirmDirective.prototype, "nzAutofocus", void 0);
    var NzPopconfirmComponent = /** @class */ (function (_super) {
        __extends(NzPopconfirmComponent, _super);
        function NzPopconfirmComponent(cdr, elementRef, directionality, document, noAnimation) {
            var _this = _super.call(this, cdr, directionality, noAnimation) || this;
            _this.elementRef = elementRef;
            _this.noAnimation = noAnimation;
            _this.nzCondition = false;
            _this.nzPopconfirmShowArrow = true;
            _this.nzOkType = 'primary';
            _this.nzAutoFocus = null;
            _this.nzOnCancel = new rxjs.Subject();
            _this.nzOnConfirm = new rxjs.Subject();
            _this._trigger = 'click';
            _this.elementFocusedBeforeModalWasOpened = null;
            _this._prefix = 'ant-popover';
            _this.document = document;
            return _this;
        }
        NzPopconfirmComponent.prototype.ngOnDestroy = function () {
            _super.prototype.ngOnDestroy.call(this);
            this.nzOnCancel.complete();
            this.nzOnConfirm.complete();
        };
        /**
         * @override
         */
        NzPopconfirmComponent.prototype.show = function () {
            if (!this.nzCondition) {
                this.capturePreviouslyFocusedElement();
                _super.prototype.show.call(this);
            }
            else {
                this.onConfirm();
            }
        };
        NzPopconfirmComponent.prototype.hide = function () {
            _super.prototype.hide.call(this);
            this.restoreFocus();
        };
        NzPopconfirmComponent.prototype.onCancel = function () {
            this.nzOnCancel.next();
            _super.prototype.hide.call(this);
        };
        NzPopconfirmComponent.prototype.onConfirm = function () {
            this.nzOnConfirm.next();
            _super.prototype.hide.call(this);
        };
        NzPopconfirmComponent.prototype.capturePreviouslyFocusedElement = function () {
            if (this.document) {
                this.elementFocusedBeforeModalWasOpened = this.document.activeElement;
            }
        };
        NzPopconfirmComponent.prototype.restoreFocus = function () {
            var toFocus = this.elementFocusedBeforeModalWasOpened;
            // We need the extra check, because IE can set the `activeElement` to null in some cases.
            if (toFocus && typeof toFocus.focus === 'function') {
                var activeElement = this.document.activeElement;
                var element = this.elementRef.nativeElement;
                if (!activeElement ||
                    activeElement === this.document.body ||
                    activeElement === element ||
                    element.contains(activeElement)) {
                    toFocus.focus();
                }
            }
        };
        return NzPopconfirmComponent;
    }(tooltip.NzToolTipComponent));
    NzPopconfirmComponent.decorators = [
        { type: core.Component, args: [{
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    encapsulation: core.ViewEncapsulation.None,
                    selector: 'nz-popconfirm',
                    exportAs: 'nzPopconfirmComponent',
                    preserveWhitespaces: false,
                    animations: [animation.zoomBigMotion],
                    template: "\n    <ng-template\n      #overlay=\"cdkConnectedOverlay\"\n      cdkConnectedOverlay\n      nzConnectedOverlay\n      [cdkConnectedOverlayHasBackdrop]=\"nzBackdrop\"\n      [cdkConnectedOverlayOrigin]=\"origin\"\n      (overlayOutsideClick)=\"onClickOutside($event)\"\n      (detach)=\"hide()\"\n      (positionChange)=\"onPositionChange($event)\"\n      [cdkConnectedOverlayPositions]=\"_positions\"\n      [cdkConnectedOverlayOpen]=\"_visible\"\n      [cdkConnectedOverlayPush]=\"true\"\n    >\n      <div\n        cdkTrapFocus\n        [cdkTrapFocusAutoCapture]=\"nzAutoFocus !== null\"\n        class=\"ant-popover\"\n        [ngClass]=\"_classMap\"\n        [class.ant-popover-rtl]=\"dir === 'rtl'\"\n        [ngStyle]=\"nzOverlayStyle\"\n        [@.disabled]=\"noAnimation?.nzNoAnimation\"\n        [nzNoAnimation]=\"noAnimation?.nzNoAnimation\"\n        [@zoomBigMotion]=\"'active'\"\n      >\n        <div class=\"ant-popover-content\">\n          <div class=\"ant-popover-arrow\" *ngIf=\"nzPopconfirmShowArrow\"></div>\n          <div class=\"ant-popover-inner\">\n            <div>\n              <div class=\"ant-popover-inner-content\">\n                <div class=\"ant-popover-message\">\n                  <ng-container *nzStringTemplateOutlet=\"nzTitle\">\n                    <ng-container *nzStringTemplateOutlet=\"nzIcon; let icon\">\n                      <i nz-icon [nzType]=\"icon || 'exclamation-circle'\" nzTheme=\"fill\"></i>\n                    </ng-container>\n                    <div class=\"ant-popover-message-title\">{{ nzTitle }}</div>\n                  </ng-container>\n                </div>\n                <div class=\"ant-popover-buttons\">\n                  <button\n                    nz-button\n                    #cancelBtn\n                    [nzSize]=\"'small'\"\n                    (click)=\"onCancel()\"\n                    [attr.cdkFocusInitial]=\"nzAutoFocus === 'cancel' || null\"\n                  >\n                    <ng-container *ngIf=\"nzCancelText\">{{ nzCancelText }}</ng-container>\n                    <ng-container *ngIf=\"!nzCancelText\">{{ 'Modal.cancelText' | nzI18n }}</ng-container>\n                  </button>\n                  <button\n                    nz-button\n                    #okBtn\n                    [nzSize]=\"'small'\"\n                    [nzType]=\"nzOkType\"\n                    (click)=\"onConfirm()\"\n                    [attr.cdkFocusInitial]=\"nzAutoFocus === 'ok' || null\"\n                  >\n                    <ng-container *ngIf=\"nzOkText\">{{ nzOkText }}</ng-container>\n                    <ng-container *ngIf=\"!nzOkText\">{{ 'Modal.okText' | nzI18n }}</ng-container>\n                  </button>\n                </div>\n              </div>\n            </div>\n          </div>\n        </div>\n      </div>\n    </ng-template>\n  "
                },] }
    ];
    NzPopconfirmComponent.ctorParameters = function () { return [
        { type: core.ChangeDetectorRef },
        { type: core.ElementRef },
        { type: bidi.Directionality, decorators: [{ type: core.Optional }] },
        { type: undefined, decorators: [{ type: core.Optional }, { type: core.Inject, args: [common.DOCUMENT,] }] },
        { type: noAnimation.NzNoAnimationDirective, decorators: [{ type: core.Host }, { type: core.Optional }] }
    ]; };
    NzPopconfirmComponent.propDecorators = {
        okBtn: [{ type: core.ViewChildren, args: ['okBtn', { read: core.ElementRef },] }],
        cancelBtn: [{ type: core.ViewChildren, args: ['cancelBtn', { read: core.ElementRef },] }]
    };

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */
    var NzPopconfirmModule = /** @class */ (function () {
        function NzPopconfirmModule() {
        }
        return NzPopconfirmModule;
    }());
    NzPopconfirmModule.decorators = [
        { type: core.NgModule, args: [{
                    declarations: [NzPopconfirmComponent, NzPopconfirmDirective],
                    exports: [NzPopconfirmComponent, NzPopconfirmDirective],
                    entryComponents: [NzPopconfirmComponent],
                    imports: [
                        bidi.BidiModule,
                        common.CommonModule,
                        button.NzButtonModule,
                        overlay.OverlayModule,
                        i18n.NzI18nModule,
                        icon.NzIconModule,
                        outlet.NzOutletModule,
                        overlay$1.NzOverlayModule,
                        noAnimation.NzNoAnimationModule,
                        tooltip.NzToolTipModule,
                        a11y.A11yModule
                    ]
                },] }
    ];

    /**
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/LICENSE
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.NzPopconfirmComponent = NzPopconfirmComponent;
    exports.NzPopconfirmDirective = NzPopconfirmDirective;
    exports.NzPopconfirmModule = NzPopconfirmModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
